test/extensions/quic_listeners/quiche/active_quic_listener_test.cc

```diff
@@ -84,102 +184,81 @@ class ActiveQuicListenerTest : public testing::TestWithParam<Network::Address::I
-
-  dispatcher_->run(Event::Dispatcher::RunType::Block);
-
-  Buffer::InstancePtr result_buffer(new Buffer::OwnedImpl());
-  const uint64_t bytes_to_read = 11;
-  uint64_t bytes_read = 0;
-  int retry = 0;
-
-  do {
-    Api::IoCallUint64Result result =
-        result_buffer->read(client_socket_->ioHandle(), bytes_to_read - bytes_read);
-
-    if (result.ok()) {
-      bytes_read += result.rc_;
-    } else if (retry == 10 || result.err_->getErrorCode() != Api::IoError::IoErrorCode::Again) {
-      break;
-    }
+  ConfigureMocks(/* connection_count = */ 1);
+  GenerateCHLO(quic::test::TestConnectionId(1));
+  dispatcher_->run(Event::Dispatcher::RunType::NonBlock);
+  ReadFromClientSockets();
+}
 
-    if (bytes_read == bytes_to_read) {
-      break;
-    }
+TEST_P(ActiveQuicListenerTest, ProcessBufferedChlos) {
+  EnvoyQuicDispatcher* const envoy_quic_dispatcher =
+      ActiveQuicListenerPeer::quic_dispatcher(*quic_listener_);
+  quic::QuicBufferedPacketStore* const buffered_packets =
+      quic::test::QuicDispatcherPeer::GetBufferedPackets(envoy_quic_dispatcher);
+
+  ConfigureMocks(ActiveQuicListener::kNumSessionsToCreatePerEpoll + 2);
+
+  // Generate one more CHLO than can be processed immediately.
+  for (size_t i = 1; i <= ActiveQuicListener::kNumSessionsToCreatePerEpoll + 1; ++i) {
+    GenerateCHLO(quic::test::TestConnectionId(i));
+  }
+  dispatcher_->run(Event::Dispatcher::RunType::NonBlock);
+
+  // The first kNumSessionsToCreatePerEpoll CHLOs are processed,
+  // the last one is buffered.
+  for (size_t i = 1; i <= ActiveQuicListener::kNumSessionsToCreatePerEpoll; ++i) {
+    EXPECT_FALSE(buffered_packets->HasBufferedPackets(quic::test::TestConnectionId(i)));
+  }
+  EXPECT_TRUE(buffered_packets->HasBufferedPackets(
+      quic::test::TestConnectionId(ActiveQuicListener::kNumSessionsToCreatePerEpoll + 1)));
+  EXPECT_TRUE(buffered_packets->HasChlosBuffered());
 
-    retry++;
-    ::usleep(10000);
-  } while (true);
-  // TearDown() will close the connection.
-  EXPECT_CALL(network_connection_callbacks_, onEvent(Network::ConnectionEvent::LocalClose));
+  // Generate more data to trigger a socket read during the next event loop.
+  GenerateCHLO(quic::test::TestConnectionId(ActiveQuicListener::kNumSessionsToCreatePerEpoll + 2));
+  dispatcher_->run(Event::Dispatcher::RunType::NonBlock);
+
+  // The socket read results in processing all CHLOs.
+  for (size_t i = 1; i <= ActiveQuicListener::kNumSessionsToCreatePerEpoll + 2; ++i) {
+    EXPECT_FALSE(buffered_packets->HasBufferedPackets(quic::test::TestConnectionId(i)));
+  }
+  EXPECT_FALSE(buffered_packets->HasChlosBuffered());
+
+  ReadFromClientSockets();
+
+  // Cancel buffered packet expiration alarm to avoid failing assertion in
+  // QuicBufferedPacketStore destructor.
+  auto alarm = quic::test::QuicBufferedPacketStorePeer::expiration_alarm(buffered_packets);
+  if (alarm) {
+    alarm->Cancel();
```

Comment:
TODO to remove this as I will relax the assertion in https://github.com/envoyproxy/envoy/pull/8496. Ditto for the dispatcher test